### PR TITLE
DEVPROD-687 Add single branch and filtered cloning

### PR DESCRIPTION
**Pull request by ZackarySantana**
DEVPROD-687

[Shrub PR](https://github.com/evergreen-ci/shrub/pull/49)

### Description
Change cloning to be more restrictive in what files it clones and what branches it clones. This effectively adds the options for `--single-branch` and `--filter=blob:0` as `FullClone` to be a parameter for the `git.get_project` command. It is intentionally a reverse parameter where the default is false, to enforce this by default on all cloning.

This option (as well as the `branch` value we have in our options) can't be that effective without some changes to how we clone. Previously, we clone the main branch, then `git reset --hard <sha>` or/and `git fetch origin <remote branch>` which means we clone most/all the project on every clone (for PR's and GH merge queue in particular). 

For PR's and GH merge queue, we previously cloned main, `git fetch origin <pull_request>`, then `git checkout/reset <sha>`. This changes both:
* For PR's, we now`git clone <fork or repo> --branch <PR branch>`
     * [Example with only single branch](https://spruce-staging.corp.mongodb.com/task/zackary_bisect_ubuntu_generator_patch_c23ade0297221b87c57173e513f5fcdf7a353c0d_661d9466862e9f0007655489_24_04_15_20_56_14/logs?execution=0&logtype=agent): Notice how it cloned from `ZackarySantana` user rather than `evergreen-ci`
     * [Example with single branch and filter](https://spruce-staging.corp.mongodb.com/task/zackary_bisect_ubuntu_generator_patch_c23ade0297221b87c57173e513f5fcdf7a353c0d_661ec0ed8592fc0008078fb3_24_04_16_18_18_28/logs?execution=0&logtype=agent)
* For GH merge queue, we now `git clone <repo> --branch <GH provided virtual branch`
     * [Example with only single branch](https://spruce-staging.corp.mongodb.com/task/zackary_bisect_ubuntu_generator_patch_7909e34c0f060ce4dc13df21dd7c32f2044d7585_661d87653e25a40007fe54b4_24_04_15_20_00_46/logs?execution=0&logtype=agent): Notice the `gh-readonly-queue/zackary_bisect/pr-687-7909e34c0f060ce4dc13df21dd7c32f2044d7585` branch is pulled, which is a branch created by github for its commit queue.
     * [Example with single branch and filter](https://spruce-staging.corp.mongodb.com/task/zackary_bisect_ubuntu_generator_patch_c23ade0297221b87c57173e513f5fcdf7a353c0d_661ec1b08592fc0008079020_24_04_16_18_21_38/logs?execution=0&logtype=agent)

### Testing
Updated existing unit tests and made the two versions (one via PR and one via GH merge queue) above. 

### Documentation
Add a note for the option `FullClone`.

### CODE REVIEW

**Comment by ZackarySantana** [comment]
Here's a comparison for time savings with "spruce" and "ui" as modules:
[With flags](https://ui.honeycomb.io/mongodb-4b/environments/staging/datasets/evergreen-agent/trace/hQpYk3uuBZG?fields%5B%5D=s_name&fields%5B%5D=s_serviceName&span=a4e8eafbb6e4c729)
Screenshot 2024-04-16 at 4 28 57 PM

[without flags](https://ui.honeycomb.io/mongodb-4b/environments/staging/datasets/evergreen-agent/trace/uSuurc9LuDu?fields%5B%5D=s_name&fields%5B%5D=s_serviceName&span=e77008755ae02fa0)
Screenshot 2024-04-16 at 4 29 46 PM

**Comment by ZackarySantana** [comment]
TBA is to add a simple fallback to previously do what we did!

File: agent/command/git_test.go
Comment (Kimchelly) [inline agent/command/git_test.go:477]: nit: remove print.

**Review by Kimchelly — CHANGES_REQUESTED** [review]
Do we also plan to add these options to shrub?

**Comment by Kimchelly** [comment]
Gonna also re-request evg-app for a second reviewer since this seems like a fairly large change to an existing feature.

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:81]: For monitoring, I want to add a fallback to the existing way (re-add it since I deleted it) and log to splunk that the new way failed- 

I've done some testing, this should not change any user-facing workflow and should only be a performance improvement but I agree that it would be nice to announce in the comms channel so there is more awareness just in case!

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:320]: Yeah it will break, I will make sure to test restarting an older task once I implement the fallback!

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:322]: GitHub promises that tokens that have read access to a private repo, has read access to (private) forks of that repo. I talked about this with Brian, and I'm going to make sure by creating a private, forkable repo and test my fork's PR.

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:252]: I'm asking runtime env. - but as a worst case scenario, the fallback should just run instead without this option so it should be backwards compatible in that regard

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:323]: [Here](https://spruce-staging.corp.mongodb.com/task/zackary_bisect_ubuntu_generator_patch_7909e34c0f060ce4dc13df21dd7c32f2044d7585_661d86843e25a40007fe5459_24_04_15_19_56_55/logs?execution=2) is an old task that I restarted. Execution #1 happened a while back, pre changes, then I restarted it without the fallback and it failed as expected, then I restarted it with the fallback and it successfully fellback.

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:252]: It looks like some older (deprecated) distros run  2.13.2 where we would need 2.27 to use --filter (but --single is from like 1.7).  In those cases, it throws an error since it's an unknown option which would bubble up to retrying with the fallback.

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go]: It looks like our token [_does not_](https://parsley-staging.corp.mongodb.com/evergreen/github_merge_queue_sandbox_ubuntu_clone_module_patch_87e52a3dfde06f7f8da356373912442cb73ef648_662170dcc7096000079ea9d9_24_04_18_19_13_34/0/all?bookmarks=0,320&shareLine=201) have access to private forks. I'm looking in to this more now, but it safely failed back on the correct repo and branch.

Comment (ZackarySantana) [reply agent/command/git.go]: This is a cool honeycomb log showing clone_source failing shortly (100ms) and the trying again.
Screenshot 2024-04-18 at 3 20 04 PM

Comment (ZackarySantana) [reply agent/command/git.go]: This is a cool honeycomb log showing clone_source passing first time (this is from a PR that was based on a branch from that private repo)
Screenshot 2024-04-18 at 3 20 59 PM

Comment (ZackarySantana) [reply agent/command/git.go]: I created a ticket with GitHub for this- they'll probably answer something!

**Comment by ZackarySantana** [comment]
[Shrub PR](https://github.com/evergreen-ci/shrub/pull/49)

Once this one is merged, I will merge that one (not earlier just in case)

File: agent/command/git.go
Comment (Kimchelly) [inline agent/command/git.go:586]: Should there be a log here? Maybe it's not relevant given the other conversation.

File: agent/command/git.go
Comment (Kimchelly) [reply agent/command/git.go:81]: A small heads up announcement SGTM!

It might be difficult to log to Splunk since this is in the agent, and only the app server can easily log to Splunk (you could do it, but it would have to be a critical or higher log).

Maybe adding a Honeycomb attribute to indicate when it's falling back to full clone would help? It's not perfect and Honeycomb wouldn't always show every full clone fallback due to sampling, but you could still see if it happens occasionally.

File: agent/command/git.go
Comment (Kimchelly) [reply agent/command/git.go]: Just falling back sounds reasonable to me! Given that there's a workaround and this is solely a performance benefit, I don't think we need to make a separate ticket to make it work.

**Review by Kimchelly — COMMENTED** [review]
The current changes all look good, just have the one question about Splunk/Honeycomb for monitoring the fallback.

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:586]: Yes that was me telling my future self to ask about how to log through to splunk-- I'm going to do the honeycomb recommendation since it would be easier and I can just query for it

Comment (ZackarySantana) [reply agent/command/git.go:586]: wait.. there's already a honeycomb attribute for attempt, so it wouldn't tell use that much actually.

**Review by Kimchelly — APPROVED** [review]
LGTM, nice job!

A second review of this PR would also be valuable!

File: agent/command/git.go
Comment (ablack12) [reply agent/command/git.go:81]: I think building a tiny route to hit the app servers and log also wouldn't be the end of the world! Honeycomb sampling might not be preferable if you're trying to find edge cases.

File: agent/command/git.go
Comment (ablack12) [inline agent/command/git.go:115]: What do you mean by this last part, "This affects GH merge queue and PR tasks"? Do other tasks not retry, or do they not use these new flags?

File: agent/command/git.go
Comment (ablack12) [inline agent/command/git.go:252]: Do you think there's user cases to enable cloning all branches but not full tree (or vice versa)? Wondering if there was a rationale for splitting it into two options or if it just seemed like a nice user experience (which it is! so no change needed, just curious)

File: agent/command/git.go
Comment (ablack12) [inline agent/command/git.go:327]: 💯 comments!

File: agent/command/git.go
Comment (ablack12) [inline agent/command/git.go:342]: Do we expect this case to be true, or are we just being careful? we could also probably just run a query on recent tasks to see if this is possible

File: agent/command/git.go
Comment (ablack12) [reply agent/command/git.go:586]: reminder to update this!

File: agent/command/git.go
Comment (ablack12) [inline agent/command/git.go:698]: Could we just pass module in since we've already looked it up, rather than passing in moduleName? (Or should we remove that lookup in fetchModuleSource?)

File: agent/command/git.go
Comment (ablack12) [inline agent/command/git.go:731]: Same thought, seems like this could've just been passed in at the top

File: agent/command/git.go
Comment (ablack12) [inline agent/command/git.go:839]: do we need to include the conf thought here, since we case for that in the code itself? PResumably this if statement is caught by attempt > 1 anyway (also do we start on attempt = 0 such that we just always want to fallback if we hit this function?)

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:252]: Since both apply a different field, it keeps the precedence of the other options and it helps with clarity- if they were combined it would either have to be a big name to explain both or it would have to just be an abstract name like `FasterClone` which they would need Evergreen specific context to interpret.

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:342]: This case is true for old PR's (older than this commit). We could make a ticket to remove it, but it's also safe to keep in because if, for some reason, the HeadBranch is missing, the task will test on the wrong branch and will misinform the user of what ran (unless they look at agent logs and see what the exact git clone command)

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go]: The attempt attribute on the span is already added!

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go]: I made it more clear, it affects other tasks but it has an exception for those two- which is what i meant to say

Comment (ablack12) [reply agent/command/git.go]: Do we know the cases where this will definitely fail where we can just set the fallback at the top? (I'd also expected the fallback to be temporary / removed after this code is merged but do we want to keep it full time instead? In which case instead of having a command fallbackToFullClone, should we just set the new flags to true?

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:378]: No, we want to try using the new flags with the clone command and only fallback on retry attempts. These blocks are the same as above but only execute when we are falling back

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go:585]: We really start at attempt = 1 since we increment in the beginning of the function. I.e. the first run of the function will have attempt = 1. Then on the first retry, attempt = 2 (causing it to fallback).

File: agent/command/git.go
Comment (ablack12) [reply agent/command/git.go]: cc @ZackarySantana I think this comment got missed -- do we need a fallback flag? Or could we just set the two flags to true? (i.e. should that be the same behavior as falling back to old logic?)

**Review by ablack12 — CHANGES_REQUESTED** [review]
Marking myself as reviewed since I still have this question:
https://github.com/evergreen-ci/evergreen/pull/7748#discussion_r1576783080

File: agent/command/git.go
Comment (ZackarySantana) [reply agent/command/git.go]: The cases in which it fallbacks now are old tasks/patches and forks of private repos. With time, eventually all recent tasks will have the info and that will be solved, and I have a thread open with GitHub support and our IT about forking private repos and what is needed for it.

The only concern then with forking private repos is that it may again need to fallback when we switch from using an oauth token to a dynamic access token. 

The cases where it will fallback because it is old tasks are [here](https://github.com/evergreen-ci/evergreen/pull/7748/files#diff-1e8d12ae20a5922169a78aaae96aab11adb323b77baf8910c3cbfba80b1d0012R330) and [here](https://github.com/evergreen-ci/evergreen/pull/7748/files#diff-1e8d12ae20a5922169a78aaae96aab11adb323b77baf8910c3cbfba80b1d0012R338). For forks, there's no special logic to detect it, although it would be mostly possible since we know all approved Evergreen orgs for repos, but it doesn't cost much to just automatically fallback from it failing the first time with no permission, and ideally it will be solved in the open threads I have.

Comment (ZackarySantana) [reply agent/command/git.go]: I did some more testing and investigating, and it should also fallback for CLI patches for now. For CLI patches, this adds a small amount of time since it makes two connections to GitHub since the clone is not the final target.

Once DEVPROD-3740 is done, we can include the Git tracking org/repo + branch in the patch. Right now we include the Evergreen tracking org/repo + branch in the patch, and then apply the Git patch changes. With the Git tracking info, we can make a direct clone and only apply uncommitted changes rather than a list of commits like we do now. There's some edge cases like if they're out of sync or haven't pushed for some reason, but that should be handled in that ticket and an appropriate fallback should apply for those.

And then once we use the Git tracking details, we should remove the fallback for CLI patches. I'll make a follow up ticket that depends on DEVPROD-3740 and add the fallback for CLI patches.

Comment (ZackarySantana) [reply agent/command/git.go]: I made: [DEVPROD-6795](https://jira.mongodb.org/browse/DEVPROD-6795)

Comment (ZackarySantana) [reply agent/command/git.go]: [This](https://spruce-staging.corp.mongodb.com/task/evg_ubuntu2204_test_db_patch_930ce9bc83f06a4ff555645dfcae46813463ea8a_6629104cfb03b80007cdf51a_24_04_24_13_59_52/logs?execution=2&logtype=all&sortBy=STATUS&sortDir=ASC) is a task showing the fallback working (in the parsely logs, you can see it doesn't use the new flags. If you look at previous executions, before the fallback, it does use the flags- and takes a little longer because it has to do a 2nd lookup/fetch)

**Review by ablack12 — COMMENTED** [review]
From convo -- 
1) Comment if statement to clarify why owner/repo/branch are set differently
2) remove fallbackToClone variable so that behavior is more consistent
3) add mainline commits specifically
4) default to previous behavior for untested requesters and make new tickets for those

**Comment by ZackarySantana** [comment]
I thought about what the options provided, and while it's possible to enable one and not the other, I think it's best to just lock both of them behind the same flag- for user and our code base simplicity (i'll rewrite the Shrub PR if this is accepted!). I kept the fallback functionality and only made this apply for the three, GH merge queue, PR's, and repotracker tasks

File: agent/command/git.go
Comment (ablack12) [inline agent/command/git.go:326]: i think its helpful to put a comment in this block here so its clear why we're doing this (an empty case statement can look like an accident)

File: docs/Project-Configuration/Project-Commands.md
Comment (ablack12) [inline docs/Project-Configuration/Project-Commands.md:548]: nit: "This option only"

**Review by ablack12 — APPROVED** [review]
I think this is a great improvement!

**Pull request merged**
